### PR TITLE
Throw exception if marshaller encounter missing/invalid associations …

### PR DESCRIPTION
**Pull request by Antoniossss**
PR addresses https://github.com/cakephp/cakephp/issues/8808

### CODE REVIEW

File: src/ORM/Marshaller.php [hunk 1]
```diff
@@ -77,6 +78,11 @@ protected function _buildPropertyMap($options)
             $assoc = $this->_table->association($key);
             if ($assoc) {
                 $map[$assoc->property()] = ['association' => $assoc] + $nested + ['associated' => []];
+            } else {
```
Comment (jippi) [inline src/ORM/Marshaller.php]: i think you can flip this logic around to be without `else`

`if (!$assoc && substr($key, 0, 1) !== "_") {`

Comment (Antoniossss) [inline src/ORM/Marshaller.php]: Is it breaking coding convention? IMHO this is more readable and no unessesery if-checks are done if first IF is fulfilled.

And to be clear: You meant removing else and add additional if, not changing the original if right?

Comment (dereuromark) [inline src/ORM/Marshaller.php]: @Antoniossss 
Its best practice to return/abort early.
Avoids nesting and else.

Comment (Antoniossss) [inline src/ORM/Marshaller.php]: Okay got it, how to do it? Cancel PR and make new one ?

Comment (ADmad) [inline src/ORM/Marshaller.php]: @Antoniossss Just make a new commit and push to your branch and this PR will be updated.

File: tests/TestCase/ORM/MarshallerTest.php [hunk 2]
```diff
@@ -1132,6 +1133,33 @@ public function testManyAssociations()
     }
 
     /**
+     * Test if exceptin is raised when called with [associated=>NonExistingAssociation]
+     * Previously such association has been simply ignored
```
Comment (dereuromark) [inline tests/TestCase/ORM/MarshallerTest.php]: Should get a `@return void` added.

File: src/ORM/Exception/MissingAssociationException.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @since         3.0.0
```
Comment (markstory) [inline src/ORM/Exception/MissingAssociationException.php]: Should b 3.3.0

File: src/ORM/Exception/MissingAssociationException.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ORM\Exception;
+
+use Cake\Core\Exception\Exception;
+
+/**
+ * Exception raised when marshaller hits nonexistend association while marshalling entity
+ */
+class MissingAssociationException extends Exception
```
Comment (markstory) [inline src/ORM/Exception/MissingAssociationException.php]: Do we need a discrete type for this? When contain fails we use `InvalidArgumentException`. Any reason we couldn't do that in this scenario as well?

File: tests/TestCase/ORM/MarshallerTest.php [hunk 5]
```diff
@@ -1132,6 +1133,33 @@ public function testManyAssociations()
...
+     */
+    public function testManyInvalidAssociation()
+    {
+        $this->expectException(MissingAssociationException::class);
```
Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: We generally use the PHPUnit annotations.

File: tests/TestCase/ORM/MarshallerTest.php [hunk 6]
```diff
@@ -1132,6 +1133,33 @@ public function testManyAssociations()
...
+            [
+                'comment' => 'Second post',
+                'user_id' => 2,
+                'user' => [
+                    'username' => 'jose',
+                ],
+            ],
+        ];
+        $marshall = new Marshaller($this->comments);
+        $result = $marshall->many($data, ['associated' => ['Users','People']]);
```
Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: You don't need `$result` here.

**Comment by markstory** [comment]
Looking good @Antoniossss !

**Comment by Antoniossss** [comment]
All comments duly noted. I made proper changes and pushed them.

**Comment by markstory** [comment]
Thanks @Antoniossss !

**Comment by Antoniossss** [comment]
@markstory Sure thing. Glad I could put my 5 cents into it :)

**Pull request merged**